c/src/lib/libbsp/arm/shared/stm32fxxxx/uart/hal-uart.c

```diff
@@ -0,0 +1,609 @@
+}uart_rx_request;
+
+static uint8_t tx_msg[NUM_PROCESSOR_NON_CONSOLE_UARTS][MAX_UART_TX_MSG_SIZE];
+static uint8_t rx_msg[NUM_PROCESSOR_NON_CONSOLE_UARTS][MAX_UART_RX_MSG_SIZE];
+
+//============================== ISR Definitions ==========================================
+static void stm32f_uart_dma_tx_isr(
+  void* argData
+)
+{
+  stm32f_base_uart_driver_entry* pUART =
+    (stm32f_base_uart_driver_entry*) argData;
+
+  HAL_DMA_IRQHandler(pUART->handle->hdmatx);
+}
+
+static void stm32f_uart_dma_rx_isr(
+  void* argData
+)
+{
+  stm32f_base_uart_driver_entry* pUART =
+    (stm32f_base_uart_driver_entry*) argData;
+
+  HAL_DMA_IRQHandler(pUART->handle->hdmarx);
+}
+
+static void stm32f_uart_isr(
+  void *arg
+)
+{
+  stm32f_base_uart_driver_entry* pUART = (stm32f_base_uart_driver_entry*) arg;
+
+  HAL_UART_IRQHandler(pUART->handle);
+}
+
+static HAL_StatusTypeDef stm32_uart_transmit(
+  UART_HandleTypeDef* hUartHandle,
+  const stm32f_uart_type uartType,
+  uint8_t * buf,
+  uint16_t len
+)
+{
+  HAL_StatusTypeDef ret = HAL_ERROR;
+
+  if ( (len > 0) && (buf != NULL) && (hUartHandle != NULL) ) {
+
+    switch ( uartType ) {
+
+    case STM32F_UART_MODE_POLLING:
+      ret = HAL_UART_Transmit(hUartHandle, buf, len, POLLED_TX_TIMEOUT_ms);
+      break;
+
+    case STM32F_UART_MODE_INT:
+      ret = HAL_UART_Transmit_IT(hUartHandle, buf, len);
+      break;
+
+    case STM32F_UART_MODE_DMA:
+      ret = HAL_UART_Transmit_DMA(hUartHandle, buf, len);
+      break;
+    }
+  }
+
+  return ret;
+}
+
+static HAL_StatusTypeDef stm32_uart_receive(
+  UART_HandleTypeDef* hUartHandle,
+  const stm32f_uart_type uartType,
+  uint8_t * buf,
+  size_t len,
+  rtems_interval timeout
+)
+{
+  HAL_StatusTypeDef ret = HAL_ERROR;
+
+  if ( (len > 0) && (buf != NULL) && (hUartHandle != NULL) ) {
+
+    switch ( uartType ) {
+
+    case STM32F_UART_MODE_POLLING:
+      ret = HAL_UART_Receive(hUartHandle, buf, (uint16_t) len, timeout);
+      break;
+
+    case STM32F_UART_MODE_INT:
+      ret = HAL_UART_Receive_IT(hUartHandle, buf, (uint16_t) len);
+      break;
+
+    case STM32F_UART_MODE_DMA:
+      ret = HAL_UART_Receive_DMA(hUartHandle, buf, (uint16_t) len);
+      break;
+    }
+  }
+
+  return ret;
+}
+
+static rtems_task stm32_uart_tx_task(
+  rtems_task_argument arg
+)
+{
+  size_t len;
+  rtems_status_code sc;
+  stm32_uart_driver_entry * pUart = (stm32_uart_driver_entry *) arg;
+
+  while ( 1 ) {
+
+    // wait until there is a TX request
+    sc = rtems_message_queue_receive(pUart->tx_msg_queue,
+      (void*) &tx_msg[pUart->instance],
+      &len,
+      RTEMS_WAIT,
+      RTEMS_NO_TIMEOUT);
+    _Assert(len <= sizeof(tx_msg));
+
+    // send tx data to UART hardware
+    if ( sc == RTEMS_SUCCESSFUL ) {
+      (void) stm32_uart_transmit(pUart->base_driver_info.handle,
+        pUart->base_driver_info.uart_mode, (uint8_t*) tx_msg[pUart->instance],
+        len);
+    }
+
+    // ensure that next TX request is only processed when the current
```

Comment:
I think that it would be better to use a transient event send form the ISR to the TX task instead of polling on a state change.